ircd/s_newconf.c

```diff
@@ -685,23 +685,53 @@ time_t
 valid_temp_time(const char *p)
 {
 	time_t result = 0;
+	long current = 0;
+	char *ch = strdup(p), *och;
 
-	while(*p)
-	{
-		if(IsDigit(*p))
-		{
-			result *= 10;
-			result += ((*p) & 0xF);
-			p++;
-		}
-		else
+	while (*ch) {
+		och = ch;
+		current = strtol(ch, &ch, 10);
+		if (errno == ERANGE)
 			return -1;
+		if (och == ch)
+			/* No numbers were given so return invalid */
+			return -1;
+		if (current) {
+			switch (*ch++) {
+				case '\0':
+					/* No unit was given so send it back as minutes */
+					if (errno == ERANGE)
+						return -1;
+					return (current * 60);
+				case 'm':
+					if (errno == ERANGE)
+						return -1;
+					result += (current * 60);
+					break;
+				case 'h':
+					if (errno == ERANGE)
+						return -1;
+					result += (current * 3600);
+					break;
+				case 'd':
+					if (errno == ERANGE)
+						return -1;
+					result += (current * 86400);
+					break;
+				case 'w':
+					if (errno == ERANGE)
+						return -1;
+					result += (current * 604800);
+					break;
+				default:
+					/* Bad time unit was given */
+					return -1;
+			}
+		}
 	}
-
 	if(result > (60 * 24 * 7 * 52))
 		result = (60 * 24 * 7 * 52);
```

Comment:
Because you're doing the minute unit multiplication before here, you need to include it in the cap test.

```suggestion
	if(result > (60 * 60 * 24 * 7 * 52))
		result = (60 * 60 * 24 * 7 * 52);
```